src/NodeIPC/Lib.hs

```diff
@@ -0,0 +1,186 @@
+opts = defaultOptions { sumEncoding = ObjectWithSingleField }
+
+instance ToJSON   MsgIn  where toEncoding = genericToEncoding opts
+instance FromJSON MsgIn  where parseJSON = genericParseJSON opts
+instance ToJSON   MsgOut where toEncoding = genericToEncoding opts
+instance FromJSON MsgOut where parseJSON = genericParseJSON opts
+
+-- | Port that is used to communicate between Cardano-node and Daedalus
+-- (e.g 8090)
+newtype Port = Port
+    { getPort :: Word16
+    } deriving Show
+
+instance Arbitrary Port where
+    arbitrary = Port <$> arbitrary
+
+data NodeIPCException
+    = NodeChannelNotFound
+    -- ^ Node channel was not found
+    | UnableToParseNodeChannel Text
+    -- ^ Unable to parse given 'Text' as File descriptor
+    | IPCException
+    -- ^ Exception thrown when there's something wrong with IPC
+
+instance Show NodeIPCException where
+    show = \case
+        NodeChannelNotFound          -> "Env variable NODE_CHANNEL_FD cannot be found"
+        UnableToParseNodeChannel err -> "Unable to parse NODE_CHANNEL_FD: " <> show err
+        IPCException                 -> "IOError has occured"
+
+instance Exception NodeIPCException
+
+-- | Acquire a Handle that can be used for IPC
+getIPCHandle :: IO Handle
+getIPCHandle = do
+    mFdstring <- liftIO $ lookupEnv "NODE_CHANNEL_FD"
+    case mFdstring of
+        Nothing -> throwM NodeChannelNotFound
+        Just fdstring -> case readEither fdstring of
+            Left err -> throwM $ UnableToParseNodeChannel $ toS err
+            Right fd -> liftIO $ fdToHandle fd
+
+-- | Start NodeJS IPC with given 'ReadHandle', 'WriteHandle' and 'Port'
+startNodeJsIPC :: (MonadIO m) => ReadHandle -> WriteHandle -> Port -> m ()
+startNodeJsIPC readHandle writeHandle port =
+    liftIO $ void $ ipcListener readHandle writeHandle port
+
+-- | Start IPC listener with given Handles and Port
+--
+-- when the listener recieves 'Ping' it will return 'Pong'.
+--
+-- If it recieves 'QueryPort', then the listener
+-- responds with 'ReplyPort' with 'Port',
+ipcListener :: forall m . (MonadIO m, MonadCatch m)
+            => ReadHandle
+            -- ^ Read handle
+            -> WriteHandle
+            -- ^ Write handle
+            -> Port
+            -- ^ Open port
+            -> m ()
+ipcListener readHndl@(ReadHandle rHndl) writeHndl@(WriteHandle wHndl) (Port port) = do
+    catches handleMsgIn [Handler handler, Handler handleMsgError]
+  where
+    handleMsgIn :: m ()
+    handleMsgIn = do
+        liftIO $ hSetNewlineMode rHndl noNewlineTranslation
+        send Started
+        msgIn <- readMessage readHndl
+        case msgIn of
+            QueryPort -> send (ReplyPort port) >> shutdown
+            Ping      -> do
+                send Pong
+                shutdown
+
+    send :: MsgOut -> m ()
+    send = sendMessage writeHndl
+
+    -- Huge catch all, fix this:
+    handler :: IOError -> m ()
+    handler err = do
+        liftIO $ when (isEOFError err) $ logError "its an eof"
```

Comment:
You can pull this liftIO before do and remove it from all these lines if you want.